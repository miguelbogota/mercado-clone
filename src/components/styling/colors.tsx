import { css, Global } from '@emotion/react';
import { kebabCase } from 'lodash';

/**
 * Helper function which infers keys and restricts values to an array with the colors tu use.
 */
const asVariable = <T extends unknown>(obj: { [K in keyof T]: string }) => obj;

/** Union type with the allowed colors of the theme. */
export type ThemeColor = keyof typeof _colors;

/** Helper function to return the value of a variable to use in css. */
export const getColor = (colorName: ThemeColor) => `var(--${kebabCase(colorName)})`;

/** Variables for the colors to use in the project. */
const _colors = asVariable({
  primaryTextColor: '#333333',
  secondaryTextColor: '#999999',
  hintTextColor: '#dadada',
  primaryBackgroundColor: '#eeeeee',
  secondaryBackgroundColor: '#ffffff',
  primaryColor: '#ffe600',
  secondaryColor: '#3483fa',
  errorColor: '#bd3835',
});

/** Global variables using css. */
const colors = (
  <Global
    styles={css`
      :root {
        // Colors
        ${Object.entries(_colors)
          .map(([key, value]) => `--${kebabCase(key)}: ${value};`)
          .join('')}
      }
    `}
  />
);

export default colors;
