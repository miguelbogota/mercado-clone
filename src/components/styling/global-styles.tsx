import { css, Global } from '@emotion/react';
import colors, { getColor } from '@app-components/colors';
import normalizeCss from '@app-components/normalize-css';

/** Default styles for the project. */
const globalStyles = (
  <>
    {normalizeCss}
    {colors}
    <Global
      styles={css`
        html,
        body,
        #__next {
          font-family: Barlow, Segoe UI, Roboto, -apple-system, Oxygen, Ubuntu, sans-serif;
          height: 100%;
          margin: 0;
          padding: 0;
          background-color: ${getColor('primaryBackgroundColor')};
          color: ${getColor('primaryTextColor')};
        }
      `}
    />
  </>
);

export default globalStyles;
