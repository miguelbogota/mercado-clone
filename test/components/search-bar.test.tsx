import SearchBar from '@app-components/search-bar';
import { getColor } from '@app-components/styling/colors';
import { render, screen } from '@testing-library/react';

jest.mock('next/router', () => require('next-router-mock'));

describe('Components - SearchBar', () => {
  it('should render correctly', () => {
    const { container } = render(<SearchBar />);
    expect(container).toMatchSnapshot();
  });

  it('should render the correct placeholder', () => {
    render(<SearchBar />);
    expect(screen.getByPlaceholderText('Nunca dejes de buscar')).toBeInTheDocument();
  });

  it('should render the correct button text', () => {
    render(<SearchBar />);
    expect(screen.getByText('search')).toBeInTheDocument();
  });

  it('should render the correct button color', () => {
    render(<SearchBar />);
    expect(screen.getByRole('button')).toHaveStyle({
      backgroundColor: getColor('primaryBackgroundColor'),
    });
  });
});
