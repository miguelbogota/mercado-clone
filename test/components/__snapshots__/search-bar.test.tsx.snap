// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - SearchBar should render correctly 1`] = `
<div>
  <nav
    class="css-tj372l"
  >
    <div
      class="css-ma7oqm"
    >
      <span
        class="css-1e6ma0l"
      >
        <a
          href="/"
          style="margin-right: 1rem; line-height: 0;"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2740%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="Mercado Libre Logo"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </span>
      <span
        class="css-1e6ma0l"
      >
        <form
          class="css-2pptk5"
        >
          <input
            class="css-fqgkxc"
            name="search"
            placeholder="Nunca dejes de buscar"
            type="text"
          />
          <button
            class="css-3pg67a"
            role="button"
            type="submit"
          >
            <span
              class="material-symbols-outlined"
            >
              search
            </span>
          </button>
        </form>
      </span>
    </div>
  </nav>
</div>
`;
